mtags/src/main/scala-2/scala/meta/internal/pc/ExtractMethodProvider.scala

```diff
@@ -0,0 +1,182 @@
+          case ValDef(_, name, tpt, _) if tpt.tpe != null =>
+            List((name, prettyType(tpt.tpe)))
+          case _ => Nil
+        }
+      )
+    }
+
+    def genName(path: List[Tree]): String = {
+      def defsOnPath(ts: List[Tree]): Set[String] = {
+        ts.flatMap(t =>
+          t match {
+            case Block(stats, expr) => defsOnPath(stats :+ expr)
+            case Template(_, _, body) => defsOnPath(body)
+            case DefDef(_, name, _, _, _, _) =>
+              Seq(name.toString)
+            case _ => Nil
+          }
+        ).toSet
+      }
+      val usedNames = defsOnPath(path)
+      if (!usedNames("newMethod")) "newMethod"
+      else {
+        var i = 0
+        while (usedNames(s"newMethod$i")) {
+          i += 1
+        }
+        s"newMethod$i"
+      }
+    }
+
+    def localRefs(ts: List[Tree]): Set[TermName] = {
+      val names = Set.newBuilder[TermName]
+      val defns = mutable.Set[TermName]()
+
+      def traverse(tree: Tree): Unit = tree match {
+        case Ident(name) =>
+          if (!defns(name.toTermName)) names += name.toTermName
+        case Select(qualifier, name) =>
+          if (!defns(name.toTermName)) names += name.toTermName
+          traverse(qualifier)
+        case ValDef(_, name, _, rhs) =>
+          defns.add(name)
+          traverse(rhs)
+        case _ => tree.children.foreach(traverse(_))
+      }
+
+      ts.map(traverse(_))
+
+      names.result()
+    }
+
+    def adjustIndent(
+        line: String,
+        newIndent: String,
+        oldIndent: Int
+    ): String = {
+      var i = 0
+      val additional = if (newIndent.indexOf("\t") != -1) "\t" else "  "
+      while ((line(i) == ' ' || line(i) == '\t') && i < oldIndent) {
+        i += 1
+      }
+      newIndent + additional + line.drop(i)
+    }
+
+    val path = pathTo(unit.body, range)
+    val edits =
+      for {
+        enclosing <- path.find(_.pos.toLSP.encloses(range))
+        extracted = extractFromBlock(enclosing)
+        head <- extracted.headOption
+        appl <- extracted.lastOption
+        shortenedPath =
+          path.takeWhile(src => defnPos.encloses(src.pos.toLSP))
+        stat = shortenedPath.lastOption.getOrElse(head)
+      } yield {
+        val noLongerAvailable = valsOnPath(shortenedPath)
+        val refsExtract = localRefs(extracted)
+        val withType =
+          noLongerAvailable.filter { case (key, _) =>
+            refsExtract.contains(key)
+          }.sorted
+        val typs = withType
+          .map { case (name, tpe) => s"$name: $tpe" }
+          .mkString(", ")
+        val newAppl = typedTreeAt(appl.pos)
+        val applType =
+          if (newAppl.tpe != null) s": ${prettyType(newAppl.tpe.widen)}" else ""
+        val applParams = withType.map(_._1).mkString(", ")
+        val name = genName(path)
+        val text = params.text()
+        val indent = stat.pos.column - (stat.pos.point - stat.pos.start) - 1
```

Comment:
Could we extract the indentation calculation to use in both Scala 2 and Scala 3?